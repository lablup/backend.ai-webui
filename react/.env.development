# disable opening browser on start
BROWSER=none
HOST=127.0.0.1
PORT=3081
