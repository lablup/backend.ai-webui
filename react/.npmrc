# Use default
node-linker=isolated
public-hoist-pattern= ['*eslint*', '*prettier*']
