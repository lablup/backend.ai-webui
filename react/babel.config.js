const path = require('path');

module.exports = {
  presets: [
    [
      '@babel/preset-env',
      {
        targets: {
          node: 'current',
        },
      },
    ],
    [
      '@babel/preset-react',
      {
        runtime: 'automatic',
      },
    ],
    '@babel/preset-typescript',
  ],
  plugins: ['@babel/plugin-syntax-import-attributes'],
  overrides: [
    {
      include: ['./src/**/*', path.resolve(__dirname, 'src/**/*')], // include only react/src folder
      plugins: [
        [
          'relay',
          {
            artifactDirectory: path.resolve(__dirname, 'src/__generated__'),
          },
        ],
      ],
    },
    {
      include: [
        '../packages/backend.ai-ui/src/**/*',
        path.resolve(__dirname, '../packages/backend.ai-ui/src/**/*'),
      ], // include only backend.ai-ui/src folder
      plugins: [
        [
          'relay',
          {
            artifactDirectory: path.resolve(
              __dirname,
              '../packages/backend.ai-ui/src/__generated__',
            ),
          },
        ],
      ],
    },
  ],
};
