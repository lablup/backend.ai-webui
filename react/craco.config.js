const HtmlWebpackPlugin = require("html-webpack-plugin");
const path = require("path");

const {
  getLoader,
  loaderByName,
  addBeforeLoader,
  addBeforeAssetModule,
  getAssetModule,
  assetModuleByName,
} = require("@craco/craco");

module.exports = {
  devServer: {
    proxy: {
      context: [
        "/src",
        "/config.toml",
        "/manifest",
        "/dist",
        "/resources",
        "/node_modules",
      ],
      target: "http://localhost:3081",
    },
    watchFiles: [
      "../index.html",
      "../config.toml",
      "../manifest/**/*",
      "../dist/**/*",
      "../resources/**/*",
    ],
  },

  webpack: {
    // When you change the this value, you might need to clear cache restart the dev server.
    // you can use `rm -rf node_modules/.cache` to clear cache.
    configure: (webpackConfig, { env, paths }) => {
      // `some.file?raw` will be treated as `asset/source`
      const { isFound, match } = getAssetModule(webpackConfig, (rule) => {
        return rule.oneOf;
      });

      if (isFound) {
        match.rule.oneOf = [
          {
            resourceQuery: /raw/,
            type: "asset/source",
          },
          ...match.rule.oneOf,
        ];
      } else {
        throw new Error("Cannot find asset module");
      }

      // For development when loading react bundle on other host, you need to set the public path to the dev server address.
      if (process.env.BUILD_TARGET === "electron") {
        webpackConfig.output.publicPath = "es6://";
      }

      // use `index.html` of original webUI` instead of using react specific one.
      const webuiIndexHtml = path.resolve(__dirname, "../index.html");
      webpackConfig.plugins = webpackConfig.plugins.map((plugin) => {
        if (plugin.constructor.name === "HtmlWebpackPlugin") {
          if (env === "development") {
            plugin = new HtmlWebpackPlugin({
              inject: true,
              template: webuiIndexHtml,
            });
          } else {
            plugin = new HtmlWebpackPlugin({
              inject: true,
              template: webuiIndexHtml,
              minify: {
                removeComments: true,
                collapseWhitespace: true,
                removeRedundantAttributes: true,
                useShortDoctype: true,
                removeEmptyAttributes: true,
                removeStyleLinkTypeAttributes: true,
                keepClosingSlash: true,
                minifyJS: true,
                minifyCSS: true,
                minifyURLs: true,
              },
            });
          }
        }

        return plugin;
      });
      paths.appHtml = webuiIndexHtml;

      return webpackConfig;
    },
  },
};
