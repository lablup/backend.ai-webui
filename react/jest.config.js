const path = require("path");
module.exports = {
  testEnvironment: "jsdom",
  clearMocks: true,
  setupFiles: ["jest-canvas-mock"],
  setupFilesAfterEnv: ["<rootDir>/src/setupTests.ts"],
  moduleNameMapper: {
    "^backend\\.ai-ui$": path.resolve(
      __dirname,
      "../packages/backend.ai-ui/src",
    ),
    "^backend\\.ai-ui/(.*)$": path.resolve(
      __dirname,
      "../packages/backend.ai-ui/src/$1",
    ),
    "\\.svg": "<rootDir>/__test__/svg.mock.js",
    "\\.(css|less|scss|sass)$": "identity-obj-proxy",
  },
  transformIgnorePatterns: [
    "node_modules/(?!(backend\\.ai-ui)/)",
    `!${path.resolve(__dirname, "../packages/backend.ai-ui/src")}`,
  ],
  transform: {
    "^.+\\.(js|jsx|ts|tsx)$": "babel-jest",
  },
  collectCoverageFrom: [
    "src/**/*.{js,jsx,ts,tsx}",
    "!src/**/*.d.ts",
    "!src/index.tsx",
    "!src/reportWebVitals.ts",
  ],
};
