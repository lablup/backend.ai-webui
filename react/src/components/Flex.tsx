import { theme } from "antd";
import React, { CSSProperties, PropsWithChildren } from "react";

interface FlexProps
  extends Omit<React.HTMLAttributes<HTMLDivElement>, "dir">,
    PropsWithChildren {
  direction?: "row" | "row-reverse" | "column" | "column-reverse";
  wrap?: "nowrap" | "wrap" | "wrap-reverse";
  justify?: "start" | "end" | "center" | "between" | "around";
  // | "evenly";
  align?: "start" | "end" | "center" | "baseline" | "stretch";
  gap?: number | "xxs" | "xs" | "sm" | "md" | "lg" | "xl" | "xxl";
}

const Flex: React.FC<FlexProps> = ({
  direction = "row",
  wrap = "nowrap",
  justify = "flex-start",
  align = "center",
  gap = 0,
  style,
  children,
  ...restProps
}) => {
  const { token } = theme.useToken();

  const transferConst = [justify, align];
  const transferConstStyle = transferConst.map((el) => {
    let tempTxt;
    switch (el) {
      case "start":
        tempTxt = "flex-start";
        break;
      case "end":
        tempTxt = "flex-end";
        break;
      case "between":
        tempTxt = "space-between";
        break;
      case "around":
        tempTxt = "space-around";
        break;
      default:
        tempTxt = el;
        break;
    }

    return tempTxt;
  });
  const flexStyle: CSSProperties = {
    display: "flex",
    flexDirection: direction,
    flexWrap: wrap,
    justifyContent: transferConstStyle[0],
    alignItems: transferConstStyle[1],
    ...style,
  };

  return (
    <div
      style={{
        alignItems: "stretch",
        backgroundColor: "transparent",
        border: "0 solid black",
        boxSizing: "border-box",
        display: "flex",
        flexBasis: "auto",
        flexDirection: "column",
        flexShrink: 0,
        listStyle: "none",
        margin: 0,
        minHeight: 0,
        minWidth: 0,
        padding: 0,
        position: "relative",
        textDecoration: "none",
        gap:
          // @ts-ignore
          typeof gap === "string" ? token["padding" + gap.toUpperCase()] : gap,
        ...flexStyle,
      }}
      {...restProps}
    >
      {children}
    </div>
  );
};

export default Flex;
