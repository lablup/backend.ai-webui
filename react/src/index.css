/* fix: match the icon size between Lucide and Ant.d icons */
.lucide {
  width: 1em !important;
  height: 1em !important;
  vertical-align: -0.125em !important;
}
