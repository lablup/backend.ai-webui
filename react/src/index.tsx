import reactToWebComponent from "./helper/react-to-webcomponent";
import React, { Suspense } from "react";
import { loadCustomThemeConfig } from "./helper/customThemeConfig";
import { Route, Routes } from "react-router-dom";

// Load custom theme config once in react/index.tsx
loadCustomThemeConfig();

const DefaultProviders = React.lazy(
  () => import("./components/DefaultProviders")
);
const Information = React.lazy(() => import("./components/Information"));
const SessionList = React.lazy(() => import("./pages/SessionListPage"));
const ServingList = React.lazy(() => import("./pages/ServingListPage"));
const RoutingList = React.lazy(() => import("./pages/RoutingListPage"));
const ResetPasswordRequired = React.lazy(
  () => import("./components/ResetPasswordRequired")
);
const StorageHostSettingPage = React.lazy(
  () => import("./pages/StorageHostSettingPage")
);
const StorageStatusPanel = React.lazy(
  () => import("./components/StorageStatusPanel")
);
const StorageStatusPanelFallback = React.lazy(() =>
  import("./components/StorageStatusPanel").then((m) => ({
    default: m.StorageStatusPanelFallback,
  }))
);
const CopyableCodeText = React.lazy(
  () => import("./components/CopyableCodeText")
);

customElements.define(
  "backend-ai-react-information",
  reactToWebComponent((props) => {
    return (
      <DefaultProviders {...props}>
        <Information />
      </DefaultProviders>
    );
  })
);

customElements.define(
  "backend-ai-react-session-list",
  reactToWebComponent((props) => {
    return (
      <DefaultProviders {...props}>
        <SessionList>{props.children}</SessionList>
      </DefaultProviders>
    );
  })
);

customElements.define(
  "backend-ai-react-serving-list",
  reactToWebComponent((props) => {
    return (
      <DefaultProviders {...props}>
        <Routes>
          <Route path="/serving" element={<ServingList />} />
          <Route path="/serving/:serviceId" element={<RoutingList />} />
        </Routes>
      </DefaultProviders>
    );
  })
);

customElements.define(
  "backend-ai-react-reset-password-required-modal",
  reactToWebComponent((props) => (
    <DefaultProviders {...props}>
      <ResetPasswordRequired />
    </DefaultProviders>
  ))
);

customElements.define(
  "backend-ai-react-storage-host-settings",
  reactToWebComponent((props) => {
    return (
      <DefaultProviders {...props}>
        <StorageHostSettingPage
          key={props.value}
          storageHostId={props.value || ""}
        />
      </DefaultProviders>
    );
  })
);

customElements.define(
  "backend-ai-react-storage-status-panel",
  reactToWebComponent((props) => {
    return (
      <DefaultProviders {...props}>
        <Suspense fallback={<StorageStatusPanelFallback />}>
          <StorageStatusPanel fetchKey={props.value || ""} />
        </Suspense>
      </DefaultProviders>
    );
  })
);

customElements.define(
  "backend-ai-react-copyable-code-text",
  reactToWebComponent((props) => {
    return (
      <DefaultProviders {...props}>
        <CopyableCodeText text={props.value || ""} />
      </DefaultProviders>
    );
  })
);
