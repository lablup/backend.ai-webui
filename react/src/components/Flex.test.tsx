import { render, screen } from "@testing-library/react";
import Flex from "./Flex";

describe("Flex", () => {
  test("default render", async () => {
    const { baseElement } = render(<Flex />);
    expect(baseElement).toMatchSnapshot();
  });
  test("render with custom props", async () => {
    const { baseElement } = render(
      <Flex
        direction="column"
        wrap="wrap-reverse"
        justify="center"
        align="start"
        gap="sm"
        style={{ backgroundColor: "blue" }}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  test("render with children", async () => {
    const { baseElement } = render(
      <Flex>
        <div data-testid="firstChildComponent">
          <h1> First Child </h1>
          <div data-testid="nestedChildComponent">
            <h1> Nested Child </h1>
          </div>
        </div>
        <div data-testid="secondChildComponent">
          <h1> Second Child </h1>
        </div>
      </Flex>
    );
    expect(screen.getByTestId("firstChildComponent")).toBeInTheDocument();
    expect(screen.getByTestId("secondChildComponent")).toBeInTheDocument();
    expect(screen.getByTestId("nestedChildComponent")).toBeInTheDocument();

    expect(baseElement).toMatchSnapshot();
  });
});
