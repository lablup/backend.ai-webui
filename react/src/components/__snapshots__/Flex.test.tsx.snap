// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex default render 1`] = `
<body>
  <div>
    <div
      style="align-items: center; background-color: transparent; border: 0px solid black; box-sizing: border-box; display: flex; flex-basis: auto; flex-direction: row; flex-shrink: 0; list-style: none; margin: 0px; min-height: 0; min-width: 0; padding: 0px; position: relative; text-decoration: none; gap: 0; flex-wrap: nowrap; justify-content: flex-start;"
    />
  </div>
</body>
`;

exports[`Flex render with children 1`] = `
<body>
  <div>
    <div
      style="align-items: center; background-color: transparent; border: 0px solid black; box-sizing: border-box; display: flex; flex-basis: auto; flex-direction: row; flex-shrink: 0; list-style: none; margin: 0px; min-height: 0; min-width: 0; padding: 0px; position: relative; text-decoration: none; gap: 0; flex-wrap: nowrap; justify-content: flex-start;"
    >
      <div
        data-testid="firstChildComponent"
      >
        <h1>
           First Child 
        </h1>
        <div
          data-testid="nestedChildComponent"
        >
          <h1>
             Nested Child 
          </h1>
        </div>
      </div>
      <div
        data-testid="secondChildComponent"
      >
        <h1>
           Second Child 
        </h1>
      </div>
    </div>
  </div>
</body>
`;

exports[`Flex render with custom props 1`] = `
<body>
  <div>
    <div
      style="align-items: flex-start; background-color: blue; border: 0px solid black; box-sizing: border-box; display: flex; flex-basis: auto; flex-direction: column; flex-shrink: 0; list-style: none; margin: 0px; min-height: 0; min-width: 0; padding: 0px; position: relative; text-decoration: none; gap: 12px; flex-wrap: wrap-reverse; justify-content: center;"
    />
  </div>
</body>
`;
